pkg/autoscaler/autoscaler.go

```diff
@@ -305,36 +300,36 @@ func (a *Autoscaler) Scale(ctx context.Context, now time.Time) (int32, bool) {
 	}
 	logger.Debugf("Current QPS: %v  Current concurrent clients: %v", totalCurrentQPS, totalCurrentConcurrency)
 
-	observedPanicPods := panicData.observedPods(now)
-	observedStableConcurrency := stableData.observedConcurrency(now)
-	observedPanicConcurrency := panicData.observedConcurrency(now)
-	observedStableConcurrencyPerPod := stableData.observedConcurrencyPerPod(now)
-	observedPanicConcurrencyPerPod := panicData.observedConcurrencyPerPod(now)
+	observedPanicPods := panicData.observedPods(now, a.spec.WindowPanic)
+	observedStableConcurrency := stableData.observedConcurrency(now, a.spec.Window)
+	observedPanicConcurrency := panicData.observedConcurrency(now, a.spec.WindowPanic)
+	observedStableConcurrencyPerPod := stableData.observedConcurrencyPerPod(now, a.spec.Window)
+	observedPanicConcurrencyPerPod := panicData.observedConcurrencyPerPod(now, a.spec.WindowPanic)
```

Comment:
It seems to me as if removing the `window` setting from the surrounding aggregation introduces room for bugs in this part of the code again. Do you think it's worth the refactoring above?

Reply:
For a given collection of data points, you can look at different windows within the data.  And that window can change.  For example, if a Revision is operating on a 60 sec window, an operator could change the PodAutoscaler to use a 30 sec window instead.  The autoscaler needs to apply the change, but it shouldn't throw away all the data.  So window becomes a parameter when asking questions about averages, etc..  There are unit tests for this.

However, I see your point about the danger of introducing bugs.  How about this?  I'll add `window` back as to the top-level aggregation struct.  And then add a method to update the window.

Reply:
I don't get the point of the refactoring here.

For one `Scale` func call, the windows are fixed. The data in `totalAggregation` are copied from the data in autoscaler. They don't drop data. 

>The autoscaler needs to apply the change, but it shouldn't throw away all the data.

What do you mean by this? It looks to me that the autoscaler still throw away all the data older than current stable window(line 288).

Reply:
👍 to yanweiguo's point. We recompute the entire window on each scale call (which we could potentially optimize). All of these objects only live for one scale cycle.

Reply:
You are totally right.  I've dropped the window refactoring.  Left the unit tests that assert we keep relevant data when the window changes.